ffi/diplomat/src/date.rs

```diff
@@ -0,0 +1,261 @@
+    use alloc::boxed::Box;
+    use alloc::sync::Arc;
+    use core::fmt::Write;
+    use diplomat_runtime::DiplomatResult;
+    use icu_calendar::types::IsoWeekday;
+    use icu_calendar::AnyCalendar;
+    use icu_calendar::{Date, Iso};
+
+    use crate::calendar::ffi::ICU4XCalendar;
+    use crate::errors::ffi::ICU4XError;
+
+    #[diplomat::enum_convert(IsoWeekday)]
+    pub enum ICU4XIsoWeekday {
+        Monday = 1,
+        Tuesday,
+        Wednesday,
+        Thursday,
+        Friday,
+        Saturday,
+        Sunday,
+    }
+    #[diplomat::opaque]
+    #[diplomat::transparent_convert]
+    /// An ICU4X Date object capable of containing a ISO-8601 date
+    #[diplomat::rust_link(icu::calendar::Date, Struct)]
+    pub struct ICU4XIsoDate(pub Date<Iso>);
+
+    impl ICU4XIsoDate {
+        /// Creates a new [`ICU4XIsoDate`] from the specified date and time.
+        #[diplomat::rust_link(icu::calendar::Date::new_iso_date, FnInStruct)]
+        pub fn try_new(
+            year: i32,
+            month: u8,
+            day: u8,
+        ) -> DiplomatResult<Box<ICU4XIsoDate>, ICU4XError> {
+            Date::new_iso_date(year, month, day)
+                .map(|dt| Box::new(ICU4XIsoDate(dt)))
+                .map_err(Into::into)
+                .into()
+        }
+
+        /// Convert this date to one in a different calendar
+        #[diplomat::rust_link(icu::calendar::Date::to_calendar, FnInStruct)]
+        pub fn to_calendar(&self, calendar: &ICU4XCalendar) -> Box<ICU4XDate> {
+            Box::new(ICU4XDate(self.0.to_calendar(calendar.0.clone())))
+        }
+
+        #[diplomat::rust_link(icu::calendar::Date::to_any, FnInStruct)]
+        pub fn to_any(&self) -> Box<ICU4XDate> {
+            Box::new(ICU4XDate(self.0.to_any().wrap_calendar_in_arc()))
+        }
+
+        /// Returns the 1-indexed day in the month for this date
+        #[diplomat::rust_link(icu::calendar::Date::day_of_month, FnInStruct)]
+        pub fn day_of_month(&self) -> u32 {
+            self.0.day_of_month().0
+        }
+
+        /// Returns the day in the week for this day
+        #[diplomat::rust_link(icu::calendar::Date::day_of_week, FnInStruct)]
+        pub fn day_of_week(&self) -> ICU4XIsoWeekday {
+            self.0.day_of_week().into()
+        }
+        /// Returns 1-indexed number of the month of this date in its year
+        ///
+        /// Note that for lunar calendars this may not lead to the same month
+        /// having the same ordinal month across years; use month_code if you care
+        /// about month identity.
+        #[diplomat::rust_link(icu::calendar::Date::month, FnInStruct)]
+        pub fn ordinal_month(&self) -> u32 {
+            self.0.month().ordinal
+        }
+
+        /// Returns the month code for this date. Typically something
+        /// like "M01", "M02", but can be more complicated for lunar calendars.
+        #[diplomat::rust_link(icu::calendar::Date::month, FnInStruct)]
+        pub fn month_code(
```

Comment:
Suggestion (optional): use just `month` instead of `ordinal_month` and `month_code` for ISO only